File: apps/web/src/features/capacity/types/capacity.ts
/**
 * Mirrors apps/api/app/schemas/capacity.py verbatim. This is the ONLY
 * source of capacity numbers in the frontend — see docs/adr/0003 and
 * CLAUDE.md §4/§10. Never compute effective/remaining/utilization/
 * over-allocation in TypeScript; only format/derive-a-status-label from
 * what these types already carry (see ../utils/presentation.ts).
 */

export interface DailyCapacity {
  date: string
  scheduled_hours: string
  unavailable_hours: string
  effective_capacity: string
  allocated_hours: string
  remaining_capacity: string
  /** null (not "0") when effective_capacity is 0 for this day — see
   * docs/domain-concepts.md#capacity-engine-phase-2. */
  utilization: string | null
  over_allocation: string
}

export interface PersonCapacity {
  person_id: string
  start_date: string
  end_date: string
  gross_capacity: string
  unavailable_hours: string
  effective_capacity: string
  allocated_hours: string
  remaining_capacity: string
  utilization: string | null
  over_allocation: string
  daily_breakdown: DailyCapacity[]
}

export interface TeamCapacity {
  team_id: string
  start_date: string
  end_date: string
  gross_capacity: string
  unavailable_hours: string
  effective_capacity: string
  allocated_hours: string
  remaining_capacity: string
  /** Weighted: sum(allocated) / sum(effective) across members — never an
   * average of member utilization percentages. See ADR 0003 decision on
   * team weighting. */
  utilization: string | null
  over_allocation: string
  members: PersonCapacity[]
}

export interface ProjectDailyDemand {
  date: string
  allocated_hours: string
}

export interface ProjectPersonDemand {
  person_id: string
  allocated_hours: string
}

export interface ProjectDemand {
  project_id: string
  start_date: string
  end_date: string
  allocated_hours: string
  allocated_people: number
  daily_breakdown: ProjectDailyDemand[]
  by_person: ProjectPersonDemand[]
}

File: apps/api/app/schemas/capacity.py
import uuid
from datetime import date
from decimal import Decimal

from pydantic import BaseModel


class DailyCapacityRead(BaseModel):
    """One day's full capacity picture. utilization is null (not 0) when
    effective_capacity is 0 — see PersonCapacityRead.utilization."""

    date: date
    scheduled_hours: Decimal
    unavailable_hours: Decimal
    effective_capacity: Decimal
    allocated_hours: Decimal
    remaining_capacity: Decimal
    utilization: Decimal | None
    over_allocation: Decimal


class PersonCapacityRead(BaseModel):
    """A person's capacity over [start_date, end_date]. gross_capacity/
    unavailable_hours/effective_capacity/allocated_hours are facts derived
    from WorkingSchedule + AvailabilityException + Allocation; remaining_capacity/
    utilization/over_allocation are further derived from those. utilization is
    null when effective_capacity is 0 (there's nothing to divide by — 0%
    would misleadingly suggest available-but-unused capacity, not "no
    capacity in this period" — see docs/domain-concepts.md)."""

    person_id: uuid.UUID
    start_date: date
    end_date: date
    gross_capacity: Decimal
    unavailable_hours: Decimal
    effective_capacity: Decimal
    allocated_hours: Decimal
    remaining_capacity: Decimal
    utilization: Decimal | None
    over_allocation: Decimal
    daily_breakdown: list[DailyCapacityRead]


class TeamCapacityRead(BaseModel):
    """Team totals are sums of member totals; utilization is WEIGHTED
    (sum(allocated) / sum(effective)), never an average of member
    utilization percentages — see docs/adr/0003-phase-2-capacity-engine.md.
    members preserves each person's own result so no individual
    over-allocation is hidden by the team aggregate."""

    team_id: uuid.UUID
    start_date: date
    end_date: date
    gross_capacity: Decimal
    unavailable_hours: Decimal
    effective_capacity: Decimal
    allocated_hours: Decimal
    remaining_capacity: Decimal
    utilization: Decimal | None
    over_allocation: Decimal
    members: list[PersonCapacityRead]


class ProjectDailyDemandRead(BaseModel):
    date: date
    allocated_hours: Decimal


class ProjectPersonDemandRead(BaseModel):
    person_id: uuid.UUID
    allocated_hours: Decimal


class ProjectDemandRead(BaseModel):
    """Allocated hours for a project over [start_date, end_date] — demand
    only, no capacity/forecasting concept (a project has no working schedule
    of its own to compare against)."""

    project_id: uuid.UUID
    start_date: date
    end_date: date
    allocated_hours: Decimal
    allocated_people: int
    daily_breakdown: list[ProjectDailyDemandRead]
    by_person: list[ProjectPersonDemandRead]
